androguard/core/apk/__init__.py

```diff
@@ -1501,33 +1503,200 @@ def is_androidtv(self) -> bool:
+        try:
+            if key_type == 'rsa':
+                asymmetric.rsa_pkcs1v15_verify(public_key, signature, signed_data, hash_algorithm=digest_algorithm)
+            elif key_type == 'dsa':
+                asymmetric.dsa_verify(public_key, signature, signed_data, hash_algorithm=digest_algorithm)
+            elif key_type == 'ec':
+                asymmetric.ecdsa_verify(public_key, signature, signed_data, hash_algorithm=digest_algorithm)
+            else:
+                raise ValueError(f"Unsupported key algorithm: {key_type}")
+            matching_certificate_verified = matching_certificate.chosen.dump()
+        except errors.SignatureError:
+            logger.info(
+                f"The public key of the certificate:{hashlib.sha256(matching_certificate.chosen.dump()).hexdigest()} is not associated with the signature!")
+        return matching_certificate_verified
 
-    def get_certificate(self, filename:str) -> x509.Certificate:
+    @staticmethod
+    def get_hash_algorithm(signer_info):
+        # Determine the hash algorithm from the SignerInfo
+        digest_algorithm = signer_info['digest_algorithm']['algorithm'].native
+        # Map the digest algorithm to a hash function
+        hash_algorithms = {
+            'md5': md5,
+            'sha1': sha1,
+            'sha256': sha256,
+            'sha224': sha224,
+            'sha384': sha384,
+            'sha512': sha512
+        }
+        if digest_algorithm not in hash_algorithms:
+            raise ValueError(f"Unsupported hash algorithm: {digest_algorithm}")
+        return digest_algorithm
+
+    @staticmethod
+    def find_certificate(signed_data_certificates, signer_info):
+        """
+        From the bag of certs, obtain the certificate referenced by the SignerInfo.
+
+        Args:
+            signed_data_certificates: List of certificates in the SignedData.
+            signer_info: SignerInfo object containing the issuer and serial number reference.
+
+        Returns:
+            The matching certificate if found, otherwise None.
+        """
+        matching_certificate = None
+        issuer_and_serial_number = signer_info['sid'].native
+        issuer = issuer_and_serial_number['issuer']
+        serial_number = issuer_and_serial_number['serial_number']
+
+        # Create an x509.Name object for the issuer in the SignerInfo
+        issuer_name = x509.Name.build(issuer)
+        issuer_str = issuer_name.human_friendly
+
+        for cert in signed_data_certificates:
+            if cert.name == 'certificate':
+                cert_native = cert.native
+                cert_issuer = cert_native['tbs_certificate']['issuer']
+                cert_serial_number = cert_native['tbs_certificate']['serial_number']
+
+                # Create an x509.Name object for the issuer in the certificate
+                cert_issuer_name = x509.Name.build(cert_issuer)
+                cert_issuer_str = cert_issuer_name.human_friendly
```

Comment:
The `x509.Name` is already available as `cert['tbs_certificate']['issuer']`.

Unfortunately, `.human_friendly` does not perform the same canonicalisation and will both accept and reject some certs incorrectly (e.g. the Java API does not change the order but it does normalise whitespace and case whereas this will sometimes change the order and not normalise whitespace and case).

Reply:
hm this makes it tricky to resolve. Any ideas?

Reply:
Well, the code for that is not trivial. But we don't need to care about the specifics of the output format, we're just interested in equality. So it might be doable to just compare the Name objects with:

* multiple AVAs sorted properly
* leading zeroes in OIDs stripped
* whitespace normalised/trimmed
* strings converted to uppercase (en-US), then lowercase (en-US), then normalized using normalization form KD

I think this is the android implementation used by v1 verification:
https://android.googlesource.com/platform/libcore/+/refs/heads/android14-release/ojluni/src/main/java/sun/security/x509/AVA.java#883

Reply:
FWIW it looks like directly comparing the `x509.Name` objects (not the `.native` dicts) uses the algorithm from RFC 4518, which is pretty close but not identical (it normalises whitespace and case *similarly* -- not sure if identically -- but uses NFKC instead of NFKD and doesn't sort AVAs). I've made an APK for which that fails even though `apksigner` verifies.

Reply:
Apart from being unsure about weird edge cases I found a bug in that comparison. Probably best not to use it :upside_down_face: